crates/sui-framework/sources/test/random.move

```diff
@@ -0,0 +1,157 @@
+    /// values are not safe to use for cryptographic purposes.
+    struct Random has store, drop {
+        state: vector<u8>,
+    }
+
+    /// Update the state of the generator and return a vector holding `DIGEST_LENGTH`
+    /// random bytes.
+    fun next_digest(random: &mut Random): vector<u8> {
+        random.state = hash::sha3_256(random.state);
+        random.state
+    }
+
+    /// Create a new pseudorandom generator with the given seed.
+    public fun new(seed: vector<u8>): Random {
+        Random { state: seed }
+    }
+
+    /// Use the given pseudorandom generator to generate a vector with l random bytes.
+    public fun next_bytes(random: &mut Random, l: u64): vector<u8> {
+        // We need ceil(l / DIGEST_LENGTH) digests to fill the array
+        let quotient = l / DIGEST_LENGTH;
+        let remainder = l - quotient * DIGEST_LENGTH;
+
+        let (i, output) = (0, vector[]);
+        while (i < quotient) {
+            vector::append(&mut output, next_digest(random));
+            i = i + 1;
+        };
+
+        // If quotient is not exact, fill the remaining bytes
+        if (remainder > 0) {
+            let (i, digest) = (0, next_digest(random));
+            while (i < remainder) {
+                vector::push_back(&mut output, *vector::borrow(&mut digest, i));
+                i = i + 1;
+            };
+        };
+
+        output
+    }
+
+    /// Use the given pseudorandom generator to generate a random `u64` integer.
+    public fun next_u64(random: &mut Random): u64 {
+        let bytes = next_digest(random);
+        bcs::peel_u64(&mut bcs::new(bytes))
+    }
+
+    /// Use the given pseudorandom generator to generate an `u64` integer in the range
+    /// [0, ..., 2^bit_length - 1].
+    fun next_u64_with_bit_length(random: &mut Random, bit_length: u8): u64 {
+        next_u64(random) >> (64 - bit_length)
+    }
+
+    /// Compute the bit length of n.
+    fun bit_length(n: u64): u8 {
+        if (n == 0) {
+            0
+        } else {
+            // Use binary search to find the bit length of n
+            let (length, mid) = (1, 32);
+            while (mid > 0) {
+                let n_mod_mid = n >> mid;
+                if (n_mod_mid > 0) {
+                    // The bit length of n is strictly larger than mid.
+                    length = length + mid;
+                    n = n_mod_mid;
+                };
+                mid = mid >> 1;
+            };
+            length
+        }
+    }
+
+    /// Use the given pseudo-random generator and a non-zero `upper_bound` to generate a
+    /// random `u64` integer in the range [0, ..., upper_bound - 1].
+    public fun next_u64_in_range(random: &mut Random, upper_bound: u64): u64 {
+        assert!(upper_bound > 0, 0);
+        let bit_length = bit_length(upper_bound);
+        let candidate = next_u64_with_bit_length(random, bit_length);
+        while (candidate >= upper_bound) {
```

Comment:
n00b question: why not just do `next_u64(random) % upper_bound`? This loop can be potentially infinite right?

Reply:
That's a great question that I actually wanted to discuss with you. My proposed solution uses rejection sampling which does, as you point out, have probabilistic running time (expected samplings < 2) but will give a truly uniform distribution for any given upper bound. Taking the modulus, as you suggest, is a lot simpler and deterministic, but will give a skewed distribution especially for large upper bounds. But if the generator is to be used only for tests, do you think that's alright? Because in that case, rejection sampling will definitely be an over-engineered solution.

Reply:
Alternatively, we can sample a random u128, take the modulus and cast to u64. This will minimise the bias but avoid the rejection sampling.

Reply:
> Taking the modulus, as you suggest, is a lot simpler and deterministic, but will give a non-uniform distribution upper bounds close to 2^64.

How non-uniform is it going to be? If it's not too bad it should be good enough for tests.

Reply:
I don't think it will be noticeable unless the upper bound is very large, say ~ 2^64 / n for some small n. But the distribution will still cover all values in the range. I'll add a note about the distribution and use the modulus operation instead.